### ISSUE TITLE
Simple training fails (?)

### ISSUE THREAD

**Issue by sorrge** [lucidrains/DALLE-pytorch#39]
Hi,

I made a toy synthetic dataset with various shapes and their text descriptions and trained DALL-E on it: https://gist.github.com/sorrge/ef0b4cbf53c3496a09596d53663655f9

The dataset is deterministic (no noise, no variation), with 9216 samples in total. Pictures are 32x32, and texts are up to 7 words. VAE is trained well w.r.t. reconstruction. I then train DALL-E on the whole dataset, so the task amounts to memorizing 9216 sequences of up to 23 tokens. I suppose that this should be easy enough for a network of this size.

However, it can't reach 100% accuracy, even after visibly converging after 500 epochs. The frequency of getting a perfect token reconstruction is around 55%.

I've investigated where the mistakes are made. It turns out that almost all errors happen in the first image token (see plots at the bottom of the notebook). Moreover, the errors are strongly correlated with the presence of masked text tokens at the end (where the text is shorter than the 7 token maximum). It never makes mistakes when the text has no masked tokens. This leads me to believe that there is something wrong with how I use masking, or there is a bug related to masking.

Can you help me to track down the problem? Also, this notebook can serve as a tutorial for DALL-E training, after the problem is figured out. It is self-contained, and the whole run takes about 4 hours.

Thanks!

**Comment by lucidrains** [comment]
@sorrge oh yes, I do believe you are correct

I've put in a fix https://github.com/lucidrains/DALLE-pytorch/commit/9bf1f0f43a5bae3ec737cabc0a04a1334e6b9a71 I've never worked with autoregressive sequences where the masked out tokens are sandwiched, thus the error

**Comment by sorrge** [comment]
Thank you, I'm going to try it. So far, I ran the same code, but without the masking. I made all texts the same length (with the same information content as before). Without masking, the accuracy is 100%, as expected.

**Comment by sorrge** [comment]
That fixed it! Now it gets 100% accuracy on the train set with masked text tokens.
Thank you!

**Comment by lucidrains** [comment]
@sorrge wahoo!
